rasterio/_io.pyx

```diff
@@ -820,13 +823,168 @@ cdef class DatasetReaderBase(DatasetBase):
         return sample_gen(self, xy, indexes)
 
 
+cdef class MemoryFileBase(object):
+    """Base for a BytesIO-like class backed by an in-memory file."""
+
+    def __init__(self, initial_bytes=b''):
+        """Map bytes to a file in an in-memory filesystem."""
+        cdef const char *path = NULL
+        cdef VSILFILE *vsi_handle = NULL
+
+        if not isinstance(initial_bytes, (bytearray, bytes)):
+            raise ValueError("Initial bytes must be type bytearray or bytes")
+
+        self.name = os.path.join('/vsimem', str(uuid.uuid4()))
+        self._pos = 0
+
+        self._initial_bytes = initial_bytes
```

Comment:
To prevent the initial bytes from being garbage collected in the following situation:

```python
data = bytes('abc')
memfile = MemoryFile(data)
data = None
```